service/http/pskInitialize_test.go

```diff
@@ -0,0 +1,118 @@
+	require.Equal(t, expCode, resp.StatusCode)
+}
+
+func initializePSK(t *testing.T, subjectID uuid.UUID, key string) {
+	request := httpgwTest.NewRequest(http.MethodGet, serviceHttp.WellKnownConfiguration, nil).Host(test.CLIENT_APPLICATION_HTTP_HOST).Build()
+	resp := httpgwTest.HTTPDo(t, request)
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+	configRespBody, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	var configResp pb.GetConfigurationResponse
+	err = protojson.Unmarshal(configRespBody, &configResp)
+	require.NoError(t, err)
+	require.False(t, configResp.GetIsInitialized())
+	require.Equal(t, configResp.GetOwner(), "00000000-0000-0000-0000-000000000000")
+	require.Equal(t, configResp.GetDeviceAuthenticationMode(), pb.GetConfigurationResponse_PRE_SHARED_KEY)
+
+	doInitializePSK(t, subjectID, key, http.StatusOK)
+
+	request = httpgwTest.NewRequest(http.MethodGet, serviceHttp.WellKnownConfiguration, nil).Host(test.CLIENT_APPLICATION_HTTP_HOST).Build()
+	resp = httpgwTest.HTTPDo(t, request)
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+	configRespBody, err = io.ReadAll(resp.Body)
+	require.NoError(t, err)
+	var configResp1 pb.GetConfigurationResponse
+	err = protojson.Unmarshal(configRespBody, &configResp1)
+	require.NoError(t, err)
+	require.True(t, configResp1.GetIsInitialized())
+	require.Equal(t, configResp1.GetOwner(), subjectID.String())
+}
+
+func setupClientApplicationForPSKInitialization(t *testing.T) func() {
+	cfg := test.MakeConfig(t)
+	cfg.Log.Level = zapcore.DebugLevel
+	cfg.APIs.HTTP.TLS.ClientCertificateRequired = false
+	cfg.Clients.Device.COAP.TLS.Authentication = device.AuthenticationPreSharedKey
+	cfg.Clients.Device.COAP.TLS.PreSharedKey.Key = ""
+	cfg.Clients.Device.COAP.TLS.PreSharedKey.SubjectUUIDStr = ""
+	shutDown := test.New(t, cfg)
+
+	return shutDown
+}
+
+func doSimpleOwn(t *testing.T, deviceID string, expCode int) {
+	request := httpgwTest.NewRequest(http.MethodPost, serviceHttp.OwnDevice, nil).Host(test.CLIENT_APPLICATION_HTTP_HOST).DeviceId(deviceID).Build()
+	resp := httpgwTest.HTTPDo(t, request)
+	require.Equal(t, expCode, resp.StatusCode)
+}
+
+func doDisown(t *testing.T, deviceID string, expCode int) {
+	request := httpgwTest.NewRequest(http.MethodPost, serviceHttp.DisownDevice, nil).Host(test.CLIENT_APPLICATION_HTTP_HOST).DeviceId(deviceID).Build()
+	resp := httpgwTest.HTTPDo(t, request)
+	require.Equal(t, expCode, resp.StatusCode)
+}
+
+func TestClientApplicationServerInitializeWithPSK(t *testing.T) {
+	dev := test.MustFindDeviceByName(test.DevsimName, []pb.GetDevicesRequest_UseMulticast{pb.GetDevicesRequest_IPV4})
+	shutDown := setupClientApplicationForPSKInitialization(t)
+	defer shutDown()
+
+	subjectID := uuid.MustParse("46178d21-d480-4e95-9bd3-6c9eefa8d9d8")
+	initializePSK(t, subjectID, "key")
+	doInitializePSK(t, subjectID, "key", http.StatusBadRequest)
+	doSimpleOwn(t, dev.GetId(), http.StatusNotFound)
+	getDevices(t, "")
+	doSimpleOwn(t, dev.GetId(), http.StatusOK)
```

Comment:
Why does the first own fail, but the second after getDevices succeeds? A timing issue?

Reply:
because cache of devices is empty, and getDevices fill it via multicast.